syntax = "proto3";

package emulation;

// Emulator grpc service is used as an controling point for external emulation
// subsystem which generate the user activity and sends it to our emulation
// network.
service Emulator {
    //
    // SendPayment is used to emulate the activity of one user sending payment
    // to another within the local network of our lightning node.
    rpc SendPayment (SendPaymentRequest) returns (SendPaymentResponse);

    //
    // OpenChannel is used to emulate that user has opened the channel with the
    // our lightning node.
    rpc OpenChannel (OpenChannelRequest) returns (OpenChannelResponse);

    //
    // CloseChannel is used to emulate that user has closed the channel with the
    // our lightning node.
    rpc CloseChannel (CloseChannelRequest) returns (CloseChannelResponse);

    //
    // SetUserConnected set user being offline or online, which means that all his
    // openned channels either could or could't be used for receiving and
    // sending payments.
    rpc SetUserConnected (SetUserConnectedRequest) returns (SetUserConnectedResponse);

    //
    // SetBlockGenDuration is used to set the time which is needed for blokc
    // to be generatedtime. This would impact channel creation, channel
    // update and channel close.
    rpc SetBlockGenDuration (SetBlockGenDurationRequest) returns (SetBlockGenDurationResponse);

    //
    // SetBlockchainFee is used to set the fee which blockchain takes for
    // making an computation, transaction creation, i.e. channel updates.
    rpc SetBlockchainFee (SetBlockchainFeeRequest) returns (SetBlockchainFeeResponse);
}

message SendPaymentRequest {
    // ID is unique identificator of the payment.
    string id = 1;

    // Sender is the id of user which sends the payment. If equals zero than we
    // believe that our lightning node is the sender.
    string sender = 2;

    // Receiver is the id of user who receive the payment. If equals zero than we
    // believe that our lightning node is the receiver.
    string receiver = 3;

    int64 amount = 4;
}

message SendPaymentResponse {
}

message OpenChannelRequest {
    string user_id = 1;
    int64 locked_by_user = 2;
}

message OpenChannelResponse {
    string channel_id = 3;
}

message CloseChannelRequest {
    string channel_id = 3;
}

message CloseChannelResponse {
}

message SetBlockGenDurationRequest {
    // Duration is the number of millisecond channel has to be in the locked
    // state after it openes or updates.
    int64 duration = 3;
}

message SetBlockGenDurationResponse {
}

message SetBlockchainFeeRequest {
    // fee is the number expressed in satoshis, which blockchain takes for
    // making an computation, transaction creation, i.e. channel updates.
    // TODO(andrew.shvv) Make it real world friendly - calculate fee
    // depending on tx size and fee per kilobyte.
    int64 fee = 3;
}

message SetBlockchainFeeResponse {
}

message SetUserConnectedRequest {
    bool is_online = 1;
    string user_id = 2;
}

message SetUserConnectedResponse {
}
